### ISSUE TITLE
Support for Ministral 3 14B

### ISSUE THREAD

**Issue by ani300** [torch-spyre/torch-spyre#696]
Supporting the HF and vLLM implementations of the Ministral 3 family of models, starting with 14B.

Model card and weight available in https://huggingface.co/mistralai/Ministral-3-14B-Base-2512

HF implementation ((https://github.com/huggingface/transformers/blob/v5.2.0/src/transformers/models/mistral3/modeling_mistral3.py) issue list:

Language model:

- [ ] TBD

Vision model:

- [ ] TBD

vLLM implementation (https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/mistral3.py) issue list:

Language model:

- [ ] TBD

Vision model:

- [ ] TBD

**Comment by RajalakshmiSR** [comment]
@ani300 https://huggingface.co/mistralai/Ministral-3-14B-Instruct-2512

**Comment by abhishekkunuru6-cmyk** [comment]
# Test Report — Ministral-3-14B-Instruct-2512 | CPU vs Spyre

**Date:** 2026-04-17  
**Environment:** PyTorch Operator-Level Tests on Spyre Accelerator  
**Summary:** Comprehensive validation of 309 operator patterns across 29 operator classes.

---

## Executive Summary

> **Total Test Cases:** 309  
> **Passed:** 177 ✅  
> **Failed:** 132 ❌  
> **Pass Rate:** 57.3%

---

## Per-Operator Detailed Results

*Single unified table showing all operators grouped by unique shape/stride combinations.*

| Operator | Shape | Stride | dtype | Patterns | Pass | Fail | Notes |
|---|---|---|---|---|---|---|---|
| `F.scaled_dot_product_attention` | Q:(1,32,1,128) K/V:(1,32,2048,128) | Q:(4096,128,128,1) K/V:(8388608,262144,128,1) | fp16 | 3 | 2 | 1 | Decode mode- Determinism ❌ |
| `F.scaled_dot_product_attention` | Q:(1,32,14,128) K/V:(1,32,2048,128) | Q:(57344,128,4096,1) K/V:(8388608,262144,128,1) | fp16 | 9 | 0 | 9 | Prefill failures ❌ |
| `F.scaled_dot_product_attention` | Q:(1,32,14,128) K/V:(1,32,2048,128) | Q:(57344,128,4096,1) K/V:(8388608,262144,128,1) | fp32 | 1 | 0 | 1 | FP32 ❌ |
| `F.scaled_dot_product_attention` | Q:(1,32,14,128) K/V:(1,32,2048,128) | Q:(57344,128,4096,1) K/V:(8388608,262144,128,1) | bf16 | 1 | 0 | 1 | BF16 ❌ |
| `torch.reshape` | (1,1,32,128) | (4096,128,128,1) | fp16 | 2 | 2 | 0 | ✅ |
| `torch.reshape` | (1,8,4,2048,128) | (2097152,262144,0,128,1) | fp16 | 2 | 2 | 0 | ✅ 0-stride |
| `torch.reshape` | (1,14,32,128) | (57344,4096,128,1) | fp16 | 2 | 2 | 0 | ✅ |
| `torch.reshape` | (1,32,1,128) | (4096,128,4096,1) | fp16 | 1 | 1 | 0 | ✅ decode (after transpose) |
| `torch.reshape` | (1,32,14,128) | (57344,128,4096,1) | fp16 | 1 | 0 | 1 | ❌ prefill (after transpose) |
| `F.silu` | (1,1,16384) | (16384,16384,1) | fp16 | 2 | 2 | 0 | ✅ decode |
| `F.silu` | (1,14,16384) | (229376,16384,1) | fp16 | 2 | 2 | 0 | ✅ prefill |
| `F.silu` | (1,16384,1) | (16384,1,16384) | fp16 | 1 | 1 | 0 | ✅ non-contig (after transpose) |
| `torch._C._log_api_usage_once` | Various | Various | fp16 | 20 | 0 | 20 | ❌ API not implemented |
| `torch.cat` | (1,14,64) + (1,14,64) | (896,1,14) + (896,1,14) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.cat` | (1,32,14,64) + (1,32,14,64) | (28672,64,2048,1) + (57344,128,4096,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.cat` | (1,8,14,64) + (1,8,14,64) | (7168,64,512,1) + (14336,128,1024,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.cat` | empty + (1,8,14,64) | — + (7168,64,512,1) | fp16 | 1 | 0 | 1 | ❌ empty tensor |
| `torch.cat` | (1,1,64) + (1,1,64) | (64,1,1) + (64,1,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.cat` | (1,32,1,64) + (1,32,1,64) | (2048,64,64,1) + (4096,128,4096,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.cat` | (1,8,1,64) + (1,8,1,64) | (512,64,64,1) + (1024,128,1024,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.cat` | empty + (1,8,1,64) | — + (512,64,64,1) | fp16 | 1 | 0 | 1 | ❌ empty tensor|
| `torch.zeros` | (1,8,2048,128) | — | fp16 | 1 | 1 | 0 | ✅ dtype specified |
| `torch.zeros` | (1,8,2048,128) | — | fp32 | 3 | 3 | 0 | ✅ list/tuple/grad |
| `torch.zeros` | (1,8,2048,128) | — | fp32 | 1 | 0 | 1 | ❌ out= param |
| `torch.add` | (1,32,14,128) + (1,32,14,128) | (57344,128,4096,1) + (57344,1792,128,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.add` | (1,8,14,128) + (1,8,14,128) | (14336,128,1024,1) + (14336,1792,128,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.add` | (1,14,5120) + (1,14,5120) | (71680,5120,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.add` | (1,32,1,128) + (1,32,1,128) | (4096,128,4096,1) + (4096,128,128,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.add` | (1,8,1,128) + (1,8,1,128) | (1024,128,1024,1) + (1024,128,128,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.add` | (1,1,5120) + (1,1,5120) | (5120,5120,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.add` | (1,14,1) + scalar | (14,1,1) | fp16 | 1 | 1 | 0 | ✅ tensor + eps |
| `torch.add` | (1,1,1) + scalar | (1,1,1) | fp16 | 1 | 1 | 0 | ✅ tensor + eps |
| `torch.add` | scalar + (14,) | (1,) | fp16 | 1 | 1 | 0 | ✅ 1 + tensor |
| `torch.add` | scalar + (1,) | (1,) | fp16 | 1 | 1 | 0 | ✅ 1 + tensor |
| `torch.add` | (1,32,14,128) + (1,32,14,128) | (57344,128,4096,1) + (57344,1792,128,1) | fp16 | 1 | 0 | 1 | ❌ alpha=0 |
| `torch.add` | (1,32,14,128) + (1,32,14,128) | (57344,128,4096,1) + (57344,1792,128,1) | fp16 | 1 | 0 | 1 | ❌ alpha=2 |
| `torch.add` | (1,14,5120) += (1,14,5120) | (71680,5120,1) | fp16 | 1 | 0 | 1 | ❌ in-place prefill |
| `torch.add` | (1,1,5120) += (1,1,5120) | (5120,5120,1) | fp16 | 1 | 0 | 1 | ❌ in-place decode |
| `torch.mul` | All 16 patterns | Various | fp16 | 16 | 16 | 0 | ✅ All pass |
| `torch.neg` | (1,32,14,64) | (57344,128,4096,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.neg` | (1,8,14,64) | (14336,128,1024,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.neg` | (1,32,1,64) | (4096,128,4096,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.neg` | (1,8,1,64) | (1024,128,1024,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.pow` | (1,14,5120) | (71680,5120,1) | fp16 | 1 | 1 | 0 | ✅ exp=2 |
| `torch.pow` | (1,1,5120) | (5120,5120,1) | fp16 | 1 | 1 | 0 | ✅ exp=2 |
| `torch.rsqrt` | (1,14,1) | (14,1,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.rsqrt` | (1,1,1) | (1,1,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.unsqueeze` | (1,14,128) | (1792,128,1) | fp16 | 1 | 1 | 0 | ✅ dim=1 |
| `torch.unsqueeze` | (14,) | (1,) | fp16 | 1 | 1 | 0 | ✅ dim=-1 |
| `torch.unsqueeze` | (1,1,128) | (128,128,1) | fp16 | 1 | 1 | 0 | ✅ dim=1 |
| `torch.unsqueeze` | (1,) | (1,) | fp16 | 1 | 1 | 0 | ✅ dim=-1 |
| `torch.floor` | (14,) and (1,) | (1,) | fp16 | 14 | 0 | 14 | ❌ All variants fail |
| `torch.log` | (14,) | (1,) | fp16 | 10 | 8 | 2 | Mostly ✅, clamp/cast ❌ |
| `torch.log` | (1,) | (1,) | fp16 | 4 | 2 | 2 | Basic ✅, clamp/cast ❌ |
| `torch.matmul` | (1,64,1) @ (1,1,1) | (64,1,1) @ (1,1,1) | fp16 | 1 | 1 | 0 | ✅ |
| `torch.matmul` | (1,64,1) @ (1,1,14) | (64,1,1) @ (14,14,1) | fp16 | 4 | 4 | 0 | ✅ All variants |
| `torch.mean` | (1,1,5120) | (5120,5120,1) | fp16 | 8 | 5 | 3 | dim=-1 ✅, global/dim0/cast ❌ |
| `torch.mean` | (1,14,5120) | (71680,5120,1) | fp16 | 8 | 6 | 2 | dim=-1/1 ✅, global/cast ❌ |
| `torch.transpose` | (1,64,1) | (64,1,1) | fp16 | 3 | 3 | 0 | ✅ All dims |
| `torch.transpose` | (1,64,14) | (896,14,1) | fp16 | 5 | 1 | 4 | Only (0,1) passes |
| `torch.transpose` | (1,1,32,128) | (4096,4096,128,1) | fp16 | 6 | 1 | 5 | Only (1,3) passes |
| `torch.transpose` | (1,1,8,128) | (1024,1024,128,1) | fp16 | 5 | 0 | 5 | ❌ All fail |
| `torch.transpose` | (1,32,1,128) | (4096,128,128,1) | fp16 | 6 | 1 | 5 | Only (2,3) passes |
| `torch.transpose` | (1,14,32,128) | (57344,4096,128,1) | fp16 | 5 | 0 | 5 | ❌ All fail |
| `torch.transpose` | (1,14,8,128) | (14336,1024,128,1) | fp16 | 5 | 0 | 5 | ❌ All fail |
| `torch.transpose` | (1,32,14,128) | (57344,1792,128,1) | fp16 | 5 | 0 | 5 | ❌ All fail |
| `torch.cos` | (1,14,128) | (1792,128,1) | fp16 | 4 | 4 | 0 | ✅ All variants |
| `torch.cos` | (1,1,128) | (128,128,1) | fp16 | 4 | 4 | 0 | ✅ All variants |
| `Tensor.contiguous()` | (1,14,32,128) | (57344,128,1792,1) | fp16 | 5 | 0 | 5 | ❌ Prefill fails |
| `Tensor.contiguous()` | (1,14,4096) | (57344,4096,1) | fp16 | 4 | 1 | 3 | Mostly ❌ |
| `Tensor.contiguous()` | (1,1,32,128) | (4096,128,128,1) | fp16 | 4 | 2 | 2 | Mixed |
| `Tensor.contiguous()` | (1,1,4096) | (4096,128,1) | fp16 | 4 | 3 | 1 | Mostly ✅ |
| `F.embedding` | indices:(1,14) | (14,1) | int64 | 4 | 3 | 1 | shape/values ✅, dtype ❌ |
| `F.embedding` | indices:(1,1) | (1,1) | int64 | 4 | 3 | 1 | shape/values ✅, dtype ❌ |
| `Tensor.__getitem__` | All 31 patterns | Various | mixed | 31 | 31 | 0 | ✅ Perfect compatibility |
| `torch.sin` | (1,14,128) | (1792,128,1) | fp16 | 2 | 2 | 0 | ✅ Basic + out= |
| `torch.sin` | (1,1,128) | (128,128,1) | fp16 | 2 | 2 | 0 | ✅ Basic + out= |
| `torch.div` | (14,) and (1,) | (1,) | int64 | 8 | 0 | 8 | ❌ All variants fail |
| `Tensor.index_copy_` | cache:(1,8,2048,128) index:(14,) src:(1,8,14,128) | cache:(2097152,262144,128,1) index:(1,) src:(14336,128,1024,1) | fp16 | 1 | 0 | 1 | ❌ Prefill |
| `Tensor.index_copy_` | cache:(1,8,2048,128) index:(1,) src:(1,8,1,128) | cache:(2097152,262144,128,1) index:(1,) src:(1024,128,128,1) | fp16 | 1 | 0 | 1 | ❌ Decode |
| `F.linear` | (1,14,5120) | (71680,5120,1) | fp16 | 2 | 2 | 0 | ✅ Prefill Q proj |
| `F.linear` | (1,14,5120) | (71680,5120,1) | fp16 | 2 | 2 | 0 | ✅ Prefill K proj |
| `F.linear` | (1,14,4096) | (57344,4096,1) | fp16 | 2 | 2 | 0 | ✅ Prefill O proj |
| `F.linear` | (1,14,5120) | (71680,5120,1) | fp16 | 2 | 2 | 0 | ✅ Prefill gate/up |
| `F.linear` | (1,14,16384) | (229376,16384,1) | fp16 | 2 | 2 | 0 | ✅ Prefill down proj |
| `F.linear` | (1,1,5120) | (71680,5120,1) | fp16 | 2 | 0 | 2 | ❌ Decode lm_head (vocab=131072) |
| `F.linear` | (1,1,5120) | (5120,5120,1) | fp16 | 2 | 2 | 0 | ✅ Decode Q proj |
| `F.linear` | (1,1,5120) | (5120,5120,1) | fp16 | 2 | 2 | 0 | ✅ Decode K proj |
| `F.linear` | (1,1,4096) | (4096,128,1) | fp16 | 2 | 2 | 0 | ✅ Decode O proj (non-standard stride) |
| `F.linear` | (1,1,5120) | (5120,5120,1) | fp16 | 2 | 2 | 0 | ✅ Decode gate/up |
| `F.linear` | (1,1,16384) | (16384,16384,1) | fp16 | 2 | 2 | 0 | ✅ Decode down proj |
| `Tensor.float()` | (1,64,1) | (64,1,1) | fp16 | 1 | 0 | 1 | ❌ fp16→fp32 |
| `Tensor.float()` | (1,1,14) | (14,14,1) | int64 | 2 | 0 | 2 | ❌ int64→fp32 |
| `Tensor.float()` | (1,1,14) | (14,14,1) | fp16 | 1 | 0 | 1 | ❌ fp16→fp32 |
| `Tensor.float()` | (1,1,1) | (1,1,1) | int64 | 1 | 0 | 1 | ❌ int64→fp32 |
| `Tensor.to(dtype)` | (1,64,1) | (64,1,1) | fp16→fp32 | 1 | 0 | 1 | ❌ Dtype conversion |
| `Tensor.to(dtype)` | (1,14,128) | (1792,128,1) | fp16→fp32 | 1 | 0 | 1 | ❌ Dtype conversion |
| `Tensor.to(dtype)` | (1,1,128) | (128,128,1) | fp16→fp32 | 1 | 0 | 1 | ❌ Dtype conversion |
| `Tensor.view()` | (1,14,4096) → (1,14,-1,128) | (57344,4096,1) | fp16 | 1 | 1 | 0 | ✅ |
| `Tensor.view()` | (1,14,1024) → (1,14,-1,128) | (14336,1024,1) | fp16 | 1 | 1 | 0 | ✅ |
| `Tensor.view()` | (1,1,4096) → (1,1,-1,128) | (4096,4096,1) | fp16 | 1 | 1 | 0 | ✅ |
| `Tensor.view()` | (1,1,1024) → (1,1,-1,128) | (1024,1024,1) | fp16 | 1 | 1 | 0 | ✅ |
| `Tensor.expand()` | (1,8,1,2048,128) → (1,8,4,2048,128) | (2097152,262144,262144,128,1) | fp16 | 4 | 4 | 0 | ✅ |
| `Tensor.expand()` | (1,64,1) → (1,-1,1) | (64,1,1) | fp16 | 1 | 0 | 1 | ❌ with clone |

---


**Report auto-generated from test logs by BOB.**
**All operators consolidated into a single unified table for easy comparison.**

cc: @ani300 @saurabh-srivastava4 @goutham-binnadi-gopala

**Comment by moriohara** [comment]
@ani300 is this still open?

**Comment by ani300** [comment]
Done through hf-adapters, with correct token generation
